File: rb/features.rb
# Obsidian SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/debug_feature'
require_relative 'feature/idempotency_feature'
require_relative 'feature/metrics_feature'
require_relative 'feature/paging_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module ObsidianFeatures
  def self.make_feature(name)
    case name
    when "base"
      ObsidianBaseFeature.new
    when "debug"
      ObsidianDebugFeature.new
    when "idempotency"
      ObsidianIdempotencyFeature.new
    when "metrics"
      ObsidianMetricsFeature.new
    when "paging"
      ObsidianPagingFeature.new
    when "ratelimit"
      ObsidianRatelimitFeature.new
    when "retry"
      ObsidianRetryFeature.new
    when "test"
      ObsidianTestFeature.new
    when "timeout"
      ObsidianTimeoutFeature.new
    else
      ObsidianBaseFeature.new
    end
  end
end

File: py/obsidian_sdk/features.py
# Obsidian SDK feature factory

from obsidian_sdk.feature.base_feature import ObsidianBaseFeature
from obsidian_sdk.feature.debug_feature import ObsidianDebugFeature
from obsidian_sdk.feature.idempotency_feature import ObsidianIdempotencyFeature
from obsidian_sdk.feature.metrics_feature import ObsidianMetricsFeature
from obsidian_sdk.feature.paging_feature import ObsidianPagingFeature
from obsidian_sdk.feature.ratelimit_feature import ObsidianRatelimitFeature
from obsidian_sdk.feature.retry_feature import ObsidianRetryFeature
from obsidian_sdk.feature.test_feature import ObsidianTestFeature
from obsidian_sdk.feature.timeout_feature import ObsidianTimeoutFeature


_FEATURES = {
    "base": lambda: ObsidianBaseFeature(),
    "debug": lambda: ObsidianDebugFeature(),
    "idempotency": lambda: ObsidianIdempotencyFeature(),
    "metrics": lambda: ObsidianMetricsFeature(),
    "paging": lambda: ObsidianPagingFeature(),
    "ratelimit": lambda: ObsidianRatelimitFeature(),
    "retry": lambda: ObsidianRetryFeature(),
    "test": lambda: ObsidianTestFeature(),
    "timeout": lambda: ObsidianTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
